File: beautifulsoup/multi-page/york/york.py

from base.Field import Field
from base.MultiPageScrapper import MultiPageScraper
from base.PageScraper import current_time_millis


class York(MultiPageScraper):
    def __init__(self, urls, university, department):
        super().__init__(urls, university, department)

    def create_json(self):
        for soup in self.soups:
            try:
                f = Field(
                          soup.select('h1.staff-title')[0].text.strip() + ', ' + soup.select('div.post-description')[0].text.strip(),
                          soup.select('div.staff-information a')[1]['href'].strip(),
                          [i.text.strip() for i in soup.select('div.staff-content ul:nth-of-type(1) li')],
                          soup.select('div.staff-information a')[0]['href'].strip())
                self.res.append(f)
            except Exception as e:
                self.exception += str(e).strip()


def extract_faculty_urls(base_url, number_of_pages=10):
    import requests
    from bs4 import BeautifulSoup
    faculty_urls = set()
    for i in range(1, number_of_pages):
        t = requests.get(base_url.format(str(i))).text
        soup = BeautifulSoup(t, 'lxml')
        for row in soup.select('div.img-wrap a'):
            faculty_urls.add(row['href'].strip())
    return faculty_urls


if __name__ == '__main__':
    t1 = current_time_millis()

    urls = extract_faculty_urls('http://eecs.lassonde.yorku.ca/community/faculty-members/page/{}/', 10)
    university = 'york'
    department = 'CS'
    York(urls, university, department)

    print('Time: {}'.format(current_time_millis() - t1))

File: jsoup/src/multi_page/york/York.java
package multi_page.york;

import base.Field;
import base.MultiPageScraper;
import multi_page.calgary.Calgary;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class York extends MultiPageScraper {

    public York(String country, String university, String department, List<String> urls) {
        super(country, university, department, urls);
    }

    @Override
    protected void createJson() {
        for (Document document : documents) {
            try {
                String name = document.select("header#title-area h1").text() +
                        document.select("header#title-area p").text();

                String link = document.select("div.staff-information ul li:nth-last-of-type(1) a").attr("href");

                Elements research = document.select("div.staff-content ul:nth-of-type(1) li");
                List<String> researchInterests = new ArrayList<>();
                for (Element research_interest : research)
                    researchInterests.add(research_interest.text().trim());

                String email = document.select("div.staff-information ul li:nth-last-of-type(2) a").attr("href");

                Field field = new Field(name, email, link, researchInterests);
                fields.add(field);
            } catch (Exception e) {
                exception.put(document.title(), e.getMessage());
            }
        }
    }

    @Override
    protected void setDir() {
        dir = "src/multi_page/york/";
    }

    public static void main(String[] args) {
        Long t1 = System.currentTimeMillis();

        String country = "Ca";
        String university = "York";
        String department = "CS";
        String facultyPageUrl = "http://eecs.lassonde.yorku.ca/community/faculty-members/page/";
        String uniBaseUrl = "";
        List<String> urls = extractFacultyUrls(facultyPageUrl, uniBaseUrl);
        new York(country, university, department, urls);

        System.out.println("Time: " + (System.currentTimeMillis() - t1));
    }

    private static List<String> extractFacultyUrls(String facultyPageUrl, String uniBaseUrl) {
        List<String> urls = new ArrayList<>();

        for (int i = 1; i < 10; i++) {

            try {
                Document soup = Jsoup.connect(facultyPageUrl + String.valueOf(i)).get();
                Elements links = soup.select("ul li.entry-item div.entry-text-wrap a");

                for (Element link : links)
                    urls.add(uniBaseUrl + link.attr("href"));
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        return urls;
    }
}
